matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/sync/DefaultSyncService.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2022 The Matrix.org Foundation C.I.C.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.matrix.android.sdk.internal.session.sync
+
+import androidx.lifecycle.LiveData
+import org.matrix.android.sdk.api.session.sync.SyncRequestState
+import org.matrix.android.sdk.api.session.sync.SyncService
+import org.matrix.android.sdk.internal.di.SessionId
+import org.matrix.android.sdk.internal.di.WorkManagerProvider
+import org.matrix.android.sdk.internal.session.sync.job.SyncThread
+import org.matrix.android.sdk.internal.session.sync.job.SyncWorker
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Provider
+
+internal class DefaultSyncService @Inject constructor(
+        @SessionId val sessionId: String,
+        private val workManagerProvider: WorkManagerProvider,
+        private val syncThreadProvider: Provider<SyncThread>,
+        private val syncTokenStore: SyncTokenStore,
+        private val syncRequestStateTracker: SyncRequestStateTracker,
+) : SyncService {
+    private var syncThread: SyncThread? = null
+
+    override fun requireBackgroundSync() {
+        SyncWorker.requireBackgroundSync(workManagerProvider, sessionId)
+    }
+
+    override fun startAutomaticBackgroundSync(timeOutInSeconds: Long, repeatDelayInSeconds: Long) {
+        SyncWorker.automaticallyBackgroundSync(workManagerProvider, sessionId, timeOutInSeconds, repeatDelayInSeconds)
+    }
+
+    override fun stopAnyBackgroundSync() {
+        SyncWorker.stopAnyBackgroundSync(workManagerProvider)
+    }
+
+    override fun startSync(fromForeground: Boolean) {
+        Timber.i("Starting sync thread")
+        // TODO How to check that now?
+        // assert(isOpen)
```

Comment:
Is it OK to not do this check anymore? See the TODO.

Reply:
For SyncService.doSync (SyncAndroidService now) this assert worked as exit condition. AssertionError was caught and handled and now it won't. Do we know how often it happens? And what will happen if we'll try to start sync without open session - will it fail before network request or will be executed? We shouldn't execute a request if there is no chance that it will result with success 🤔

Reply:
You are right. Improved in a33f7d0312bdd8bf8769b03aae4973797b35c124 which restores the previous checks.